### PR TITLE
CDAP-14932. Salesforce Batch Source to fetch objects via SOQL. Classes to work with Salesforce API and to parse SOQL, create schema

### PR DESCRIPTION
**Pull request by aonischuk**
PR contains classes to:
1. Create jobs to run SOQL query using Salesforce Bulk API
2. Parse out fieldNames from SOQL query
SELECT Id,NAME FROM Opportunity => ["Id", "Name"]
3. Parse out sObject name from SOQL query
SELECT Id,NAME FROM Opportunity => "Opportunity"
4. Classes which automatically generate CDAP schema using SOQL query. Which works like this:
a) Parse out sObject Name from SOQL query
b) Parse out Field Names from SOQL query
c) From Salesforce SOAP API obtain information about types of fields for given sObject
d) Using the information generate CDAP Schema
5. Unit tests

### CODE REVIEW

File: src/main/java/co/cask/hydrator/salesforce/SalesforceBulkUtil.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    job.setConcurrencyMode(ConcurrencyMode.Parallel);
+    job.setContentType(ContentType.CSV);
+    job = bulkConnection.createJob(job);
+    Preconditions.checkState(job.getId() != null, "Couldn't get job ID. Perhaps there was a problem in creating the " +
+      "batch job");
+    return bulkConnection.getJobStatus(job.getId());
+  }
+
+  /**
+   * Start batch job for reading a given guery result.
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/SalesforceBulkUtil.java]: s/for/of

File: src/main/java/co/cask/hydrator/salesforce/SalesforceBulkUtil.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      bulkConnection.createBatchFromStream(job, bout);
+    }
+
+    BatchInfo[] batches = bulkConnection.getBatchInfoList(job.getId()).getBatchInfo();
+
+    return batches;
+  }
+
+  /**
+   * Wait while a batch with given batchId succeeds, or throw an exception
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/SalesforceBulkUtil.java]: s/while/until

File: src/main/java/co/cask/hydrator/salesforce/SalesforceSchemaUtil.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+      Schema.Field schemaField = Schema.Field.of(fieldName, fieldSchema);
+      schemaFields.add(schemaField);
+    }
+
+    return Schema.recordOf("output",  schemaFields);
+  }
+
+  /**
+   * If this is relationship query (join equivalent for soql)
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/SalesforceSchemaUtil.java]: s/is/is a

File: src/main/java/co/cask/hydrator/salesforce/SalesforceSchemaUtil.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+   * Relationship field - Account.Name
+   * Nested field - (SELECT Id FROM Opportunity)
+   * Aggregate field - AVG(Amount), COUNT()
+   *
+   *
+   * @param fieldName name of the field
+   * @return is field complex (relationship or aggregate or nested)
+   */
+  private static boolean isFieldComplex(String fieldName) {
+    return (fieldName.contains("(") || fieldName.contains("."));
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/SalesforceSchemaUtil.java]: is it possible to have something like (SELECT Id FROM Opportunity) without it being nested?

File: src/main/java/co/cask/hydrator/salesforce/parser/ThrowingErrorListener.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import org.antlr.v4.runtime.Recognizer;
+
+/**
+ * Error listener which throws exception when parsing error happens, instead of just default writing to stderr.
+ */
+class ThrowingErrorListener extends BaseErrorListener {
+  @Override
+  public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol,
+                          int line, int charPositionInLine, String msg, RecognitionException e) {
+    throw new SOQLParsingException("Failed to parse query at " + line + ":" + charPositionInLine + " " + msg);
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/parser/ThrowingErrorListener.java]: can we also pass the original exception e through?

File: src/test/java/co/cask/hydrator/salesforce/parser/SalesforceQueryParserTest.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class SalesforceQueryParserTest {
+  private static final Map<String, Object[]> queries = new HashMap<>();
+
+  static {
+    queries.put("SELECT Id, Name FROM Opportunity",
```
Comment (jayfjin) [inline src/test/java/co/cask/hydrator/salesforce/parser/SalesforceQueryParserTest.java]: what happens when the following test is run: "(SELECT Id, Name FROM Opportunity)"? I believe this is a valid sql statement?

File: src/main/antlr4/soql/SOQL.g4 [hunk 7]
```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright 2017 StreamSets Inc.
```
[the remark this answers is not in the document]
Comment (jayfjin) [reply src/main/antlr4/soql/SOQL.g4]: I'll ping Bhooshan on this

File: src/test/java/co/cask/hydrator/salesforce/parser/SalesforceQueryParserTest.java (hunk 6 above)
Comment (aonischuk) [reply src/test/java/co/cask/hydrator/salesforce/parser/SalesforceQueryParserTest.java]: I agree. But salesforce server SOQL parser does not like your idea :). The bellow exception is a response from Salesforce Bulk API not from my code.

> java.lang.Exception: co.cask.hydrator.salesforce.BulkAPIBatchException: Batch failed. BatchId='7511i000002aj0uAAA', Reason='InvalidBatch : Failed to process query: MALFORMED_QUERY: unexpected token: ('
	at org.apache.hadoop.mapred.LocalJobRunner$Job.runTasks(LocalJobRunner.java:489) ~[org.apache.hadoop.hadoop-mapreduce-client-common-2.8.0.jar:na]
	at org.apache.hadoop.mapred.LocalJobRunner$Job.run(LocalJobRunner.java:549) ~[org.apache.hadoop.hadoop-mapreduce-client-common-2.8.0.jar:na]
Caused by: co.cask.hydrator.salesforce.BulkAPIBatchException: Batch failed. BatchId='7511i000002aj0uAAA', Reason='InvalidBatch : Failed to process query: MALFORMED_QUERY: unexpected token: ('

File: src/main/java/co/cask/hydrator/salesforce/SalesforceSchemaUtil.java (hunk 4 above)
Comment (aonischuk) [reply src/main/java/co/cask/hydrator/salesforce/SalesforceSchemaUtil.java]: Salesforce Bulk API responds with an error to "(SELECT Id FROM Opportunity)" query.
> BulkAPIBatchException: Batch failed. BatchId='7511i000002aj0uAAA', Reason='InvalidBatch : Failed to process query: MALFORMED_QUERY: unexpected token: ('

File: src/main/antlr4/soql/SOQL.g4 (hunk 7 above)
[the remark this answers is not in the document]
Comment (aonischuk) [reply src/main/antlr4/soql/SOQL.g4]: Also I note I didn't find this lib in maven repos.

**Comment by aonischuk** [comment]
@jayfjin  addressed all the comments. Can we merge this in at this point?

File: src/test/java/co/cask/hydrator/salesforce/parser/SalesforceQueryParserTest.java (hunk 6 above)
Comment (jayfjin) [reply src/test/java/co/cask/hydrator/salesforce/parser/SalesforceQueryParserTest.java]: Hmm ok, thanks for testing

Comment (aonischuk) [reply src/test/java/co/cask/hydrator/salesforce/parser/SalesforceQueryParserTest.java]: thanks for reviewing

**Pull request merged**